java/arcs/core/storage/ReferenceModeStore.kt

```diff
@@ -183,25 +173,17 @@ class ReferenceModeStore private constructor(
     override suspend fun idle() {
         backingStore.idle()
         containerStore.idle()
+        receiveQueue.flush().join()
     }
 
     override fun on(
         callback: ProxyCallback<RefModeStoreData, RefModeStoreOp, RefModeStoreOutput>
     ): Int = callbacks.register(callback)
 
-    override fun off(callbackToken: Int) {
-        callbacks.unregister(callbackToken)
-        if (callbacks.isEmpty() && !callbacksStateChannel.isClosedForSend) {
-            try {
-                callbacksStateChannel.offer(false)
-            } catch (e: ClosedSendChannelException) {
-                // No-op. If the channel is closed (which can happen between the if's check and the
-                // offer call above), then it's no big deal.
-                log.debug {
-                    "Attempted to send false to the callbacksStateChannel when it was already " +
-                        "closed."
-                }
-            }
+    override suspend fun off(callbackToken: Int) {
+        if (callbacks.unregister(callbackToken) == 0) {
+            callbacksHasBecomeEmpty = true
+            receiveQueue.flush()
```

Comment:
this is a bit strange. how about instead you launch cleanupIfNeeded then return, and inside cleanupIfNeeded you .flush().join()
I think in that way you can also void the suspend here

Reply:
I don't think it's possible to avoid the suspend here: `receiveQueue.flush` may result in triggering the cleanup function, which is a suspending function, so we need to be in a coroutine context.